src/Parser.y

```diff
@@ -418,8 +418,40 @@ INSTR :: { G.Instruction }
   | ifBegin IFCASES ifEnd                                               { G.InstIf (reverse $2) }
   | switchBegin EXPR colon SWITCHCASES DEFAULTCASE switchEnd            { G.InstSwitch $2 (reverse ($5 : $4)) }
   | switchBegin EXPR colon SWITCHCASES switchEnd                        { G.InstSwitch $2 (reverse $4) }
-  | forBegin ID with EXPR souls untilLevel EXPR CODEBLOCK forEnd        { G.InstFor $2 $4 $7 $8 }
-  | forEachBegin ID withTitaniteFrom EXPR CODEBLOCK forEachEnd          { G.InstForEach $2 $4 $5 }
+  | FORBEGIN with EXPR souls untilLevel EXPR CODEBLOCK forEnd           {% do
+                                                                          checkIterVariables $7
+                                                                          (ST.SymTable d s cs (_:ivs)) <- RWS.get
+                                                                          RWS.put (ST.SymTable d s cs ivs)
+                                                                          return $ G.InstFor $1 $3 $6 $7 }
+  | FOREACHBEGIN withTitaniteFrom EXPR CODEBLOCK forEachEnd             {% do
+                                                                          checkIterVariables $4
+                                                                          (ST.SymTable d s cs (_:ivs)) <- RWS.get
+                                                                          RWS.put (ST.SymTable d s cs ivs)
+                                                                          return $ G.InstForEach $1 $3 $4 }
+
+FORBEGIN :: { G.Id }
+  : forBegin ID                                                         {% do
+                                                                          mid <- checkIdAvailability $2
+                                                                          (ST.SymTable d s cs ivs) <- RWS.get
+                                                                          case mid of
+                                                                            Just ST.DictionaryEntry {ST.name=varName} -> do
+                                                                              RWS.put (ST.SymTable d s cs (varName:ivs))
+                                                                            Nothing -> do
+                                                                              -- checkIdAvailability already rose an error here, nothing else to do
+                                                                              RWS.put (ST.SymTable d s cs ("_UNDECLARED_ITER_VAR":ivs))
```

Comment:
why do we add `"_UNDECLARED_ITER_VAR"` here?

Reply:
removed